source/ExclusiveItem.h

```diff
@@ -0,0 +1,50 @@
+/* ExclusiveItem.h
+Copyright (c) 2022 by Amazinite
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef EXCLUSIVE_ITEM_H_
+#define EXCLUSIVE_ITEM_H_
+
+#include <memory>
+#include <utility>
+
+
+
+// An ExclusiveItem is an object stored as either a pointer or a stock
+// object, but not both.
+template <class Type>
+class ExclusiveItem {
+public:
+	ExclusiveItem() = default;
+
+	ExclusiveItem(const Type *item) : stockItem(item) {}
+	explicit ExclusiveItem(Type &&item) : item(std::move(item)) {}
+
+	ExclusiveItem(ExclusiveItem<Type> &&other) : stockItem(other.stockItem), item(std::move(other.item)) {}
+	ExclusiveItem(const ExclusiveItem<Type> &other) : stockItem(other.stockItem), item(other.item) {}
+
+	const Type *operator->() const noexcept { return stockItem ? stockItem : std::addressof(item); }
+	const Type &operator*() const noexcept { return stockItem ? *stockItem : item; }
+
+	ExclusiveItem<Type> &operator=(ExclusiveItem<Type> &&other) { stockItem = other.stockItem; item = std::move(other.item); return *this; }
+	ExclusiveItem<Type> &operator=(const ExclusiveItem<Type> &other) { stockItem = other.stockItem; item = other.item; return *this; }
+	bool operator==(const ExclusiveItem<Type> &other) const { return *this == *other; }
```

Comment:
Did you mean to write `**this`? I'm not too sure what you wrote will work.

Reply:
These are using the custom * operator

Reply:
But you want to use the custom * operator on both objects no?

Reply:
Is that not what's already happening? If `*this` was returning a reference to `this` instead of a reference to `item` then we'd be comparing `ExclusiveItem<Type>` to `Type`, which has no `==` operator, so I'd think it'd throw an error if that's what was happening.

Admittedly I'm not sure which * operator is being used here. At first I had `(stockItem ? *stockItem : item) == (other.stockItem ? *other.stockItem : other.item)` but that's just what the custom * operator does, so I replaced that with what's there now.

Reply:
> I'd think it'd throw an error if that's what was happening.

Yeah I think so too, you're comparing an ExclusiveItem with Type.

(Did you try to use the equality operator? It won't get instantiated otherwise)